File: benchmarks-haskell/scott-tree-map-and-sum.hs
{-# LANGUAGE LambdaCase #-}

import Interpreter

scottLeaf :: Term
scottLeaf = Lambda (Lambda (Lambda (Apply (TVar 1) (TVar 2))))

scottNode :: Term
scottNode =
  Lambda (Lambda (Lambda (Lambda (Apply (Apply (TVar 0) (TVar 3)) (TVar 2)))))

add :: Term
add = BinaryOp (+)

multiply :: Term
multiply = BinaryOp (*)

scottTreeSum :: Term
scottTreeSum =
  Fix
    ( Lambda
        ( Lambda
            ( Apply
                (Apply (TVar 0) (Lambda (TVar 0)))
                ( Lambda
                    ( Lambda
                        (Apply (Apply add (Apply (TVar 3) (TVar 1))) (Apply (TVar 3) (TVar 0)))
                    )
                )
            )
        )
    )

scottTreeMap :: Term
scottTreeMap =
  Fix
    ( Lambda
        ( Lambda
            ( Lambda
                ( Apply
                    (Apply (TVar 0) (Lambda (Apply scottLeaf (Apply (TVar 2) (TVar 0)))))
                    ( Lambda
                        ( Lambda
                            ( Apply
                                (Apply scottNode (Apply (Apply (TVar 4) (TVar 3)) (TVar 1)))
                                (Apply (Apply (TVar 4) (TVar 3)) (TVar 0))
                            )
                        )
                    )
                )
            )
        )
    )

doubleFunction :: Term
doubleFunction = Lambda (Apply (Apply multiply (TVar 0)) (Const 2))

generateTree :: Int -> Term
generateTree = \case
  1 -> Apply scottLeaf (Const 1)
  n ->
    Apply
      (Apply scottNode (generateTree (n `div` 2)))
      (generateTree (n `div` 2))

benchmarkTerm :: Term
benchmarkTerm =
  Apply
    scottTreeSum
    (Apply (Apply scottTreeMap doubleFunction) (generateTree (65536 * 2 * 2)))

main :: IO ()
main = case denote [] benchmarkTerm of
  VConst n -> print n
  _ -> error "Expected a constant result!"

File: benchmarks-ocaml/scott_tree_map_and_sum.ml
open Interpreter

let scott_leaf = Lambda (Lambda (Lambda (Apply (tvar 1, tvar 2))))

let scott_node =
  Lambda (Lambda (Lambda (Lambda (Apply (Apply (tvar 0, tvar 3), tvar 2)))))

let add = BinaryOp ( + )

let multiply = BinaryOp ( * )

let scott_tree_sum =
  Fix
    (Lambda
       (Lambda
          (Apply
             ( Apply (tvar 0, Lambda (tvar 0)),
               Lambda
                 (Lambda
                    (Apply
                       ( Apply (add, Apply (tvar 3, tvar 1)),
                         Apply (tvar 3, tvar 0) ))) ))))

let scott_tree_map =
  Fix
    (Lambda
       (Lambda
          (Lambda
             (Apply
                ( Apply
                    (tvar 0, Lambda (Apply (scott_leaf, Apply (tvar 2, tvar 0)))),
                  Lambda
                    (Lambda
                       (Apply
                          ( Apply
                              ( scott_node,
                                Apply (Apply (tvar 4, tvar 3), tvar 1) ),
                            Apply (Apply (tvar 4, tvar 3), tvar 0) ))) )))))

let double_function = Lambda (Apply (Apply (multiply, tvar 0), Const 2))

let rec generate_tree = function
  | 1 -> Apply (scott_leaf, Const 1)
  | n -> Apply (Apply (scott_node, generate_tree (n / 2)), generate_tree (n / 2))

let benchmark_term =
  Apply
    ( scott_tree_sum,
      Apply (Apply (scott_tree_map, double_function), generate_tree (65536 * 2 * 2)) )

let () =
  match denote [] benchmark_term with
  | VConst n -> Printf.printf "%d\n" n
  | _ -> failwith "Expected a constant result!"
